File: docs/pseudo-code/sse-chat-parsing/SSEChatParser.kt
/**
 * P8FS SSE Chat Parsing Flow - Android Implementation
 * 
 * Based on reference implementation from T1 app
 * Key features:
 * 1. Session ID management for conversation threads
 * 2. Comprehensive header management from content_headers.md
 * 3. SSE streaming with real-time chunk parsing
 * 4. Only send latest message, server manages thread context
 * 5. Support for audio chat via X-CHAT-IS-AUDIO header
 */

import android.util.Log
import io.ktor.client.*
import io.ktor.client.plugins.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.utils.io.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.Json
import java.util.*

data class ChatMessage(
    val role: MessageRole,
    val content: String
)

enum class MessageRole {
    USER, ASSISTANT
}

data class ChatCompletionRequest(
    val model: String = "gpt-4.1-mini",
    val messages: List<ChatMessage>,
    val temperature: Double = 0.7,
    val maxTokens: Int = 1000,
    val stream: Boolean = true
)

data class ChatCompletionChunk(
    val id: String,
    val `object`: String = "chat.completion.chunk",
    val choices: List<StreamChoice>
)

data class StreamChoice(
    val index: Int = 0,
    val delta: Delta,
    val finishReason: String? = null
)

data class Delta(
    val content: String? = null
)

enum class P8Agent(val key: String) {
    P8_SIMULATION("p8-sim"),
    P8_RESEARCH("p8-research"),
    P8_ANALYSIS("p8-analysis"),
    P8_QA("p8-qa")
}

sealed class ChatStreamEvent {
    data class Content(val text: String) : ChatStreamEvent()
    data class Error(val exception: Exception) : ChatStreamEvent()
    object Done : ChatStreamEvent()
}

/**
 * Main chat client with session management and comprehensive headers
 */
class P8FSChatClient(
    private val headersManager: HeadersManager
) {
    companion object {
        private const val TAG = "P8FSChatClient"
    }
    
    // Generate unique session ID for this chat conversation
    private val sessionId = UUID.randomUUID().toString().also { newSessionId ->
        Log.d(TAG, "🆔 Generated new chat session ID: $newSessionId")
    }
    
    private val httpClient = HttpClient {
        install(HttpTimeout) {
            requestTimeoutMillis = 60000 // 60 seconds for streaming
            connectTimeoutMillis = 10000
            socketTimeoutMillis = 60000
        }
    }
    
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }
    
    /**
     * Stream chat completion with comprehensive P8FS headers
     * Only sends the latest user message - server manages thread via session ID
     */
    fun streamChatCompletion(
        userMessage: String,
        agent: P8Agent? = P8Agent.P8_SIMULATION,
        model: String = "gpt-4.1-mini"
    ): Flow<ChatStreamEvent> = flow {
        try {
            // Create request with only the new user message
            val request = ChatCompletionRequest(
                model = model,
                messages = listOf(ChatMessage(MessageRole.USER, userMessage)),
                stream = true
            )
            
            // Determine endpoint based on agent
            val endpoint = when (agent) {
                null -> "${AppConfig.baseUrl}/api/v1/chat/completions"
                else -> "${AppConfig.baseUrl}/api/v1/agent/${agent.key}/chat/completions"
            }
            
            Log.d(TAG, "🚀 Starting streaming chat to: $endpoint")
            Log.d(TAG, "📝 User message: $userMessage")
            Log.d(TAG, "🤖 Agent: ${agent?.key ?: "none"}")
            
            // Get comprehensive headers including session ID
            val chatHeaders = headersManager.getChatHeaders(sessionId)
            Log.d(TAG, "🆔 Session ID: $sessionId")
            Log.d(TAG, "📋 Headers: ${chatHeaders.keys.joinToString(", ")}")
            
            httpClient.preparePost(endpoint) {
                // Apply all P8FS content headers
                chatHeaders.forEach { (key, value) ->
                    header(key, value)
                }
                contentType(ContentType.Application.Json)
                setBody(request)
            }.execute { response ->
                
                if (response.status != HttpStatusCode.OK) {
                    val errorText = response.bodyAsText()
                    Log.e(TAG, "❌ Streaming failed: ${response.status} - $errorText")
                    emit(ChatStreamEvent.Error(Exception("Chat failed: ${response.status}")))
                    return@execute
                }
                
                Log.d(TAG, "✅ SSE stream connected")
                
                // Parse SSE stream
                val channel = response.bodyAsChannel()
                val contentBuffer = StringBuilder()
                var doneEmitted = false
                
                while (!channel.isClosedForRead) {
                    val line = channel.readUTF8Line() ?: break
                    
                    // SSE format: "data: {json}" or "data: [DONE]"
                    if (line.startsWith("data: ")) {
                        val data = line.substring(6).trim()
                        
                        when (data) {
                            "[DONE]" -> {
                                Log.d(TAG, "🏁 Stream completed with [DONE] signal")
                                emit(ChatStreamEvent.Done)
                                doneEmitted = true
                                break
                            }
                            else -> {
                                try {
                                    val chunk = json.decodeFromString<ChatCompletionChunk>(data)
                                    
                                    // Extract content from delta
                                    chunk.choices.firstOrNull()?.delta?.content?.let { content ->
                                        Log.d(TAG, "📝 Received chunk: '$content'")
                                        emit(ChatStreamEvent.Content(content))
                                        contentBuffer.append(content)
                                    }
                                    
                                    // Check for completion
                                    chunk.choices.firstOrNull()?.finishReason?.let { reason ->
                                        Log.d(TAG, "🏁 Finish reason: $reason")
                                    }
                                    
                                } catch (e: Exception) {
                                    Log.e(TAG, "❌ Error parsing chunk: $data", e)
                                }
                            }
                        }
                    }
                }
                
                Log.d(TAG, "📊 Stream ended. Total content: ${contentBuffer.length} chars")
                
                // Ensure Done is emitted even if server doesn't send [DONE]
                if (!doneEmitted) {
                    Log.d(TAG, "🏁 Emitting Done (stream ended without [DONE])")
                    emit(ChatStreamEvent.Done)
                }
            }
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ Streaming error", e)
            emit(ChatStreamEvent.Error(e))
        }
    }
    
    /**
     * Stream chat with audio input
     * Sets X-CHAT-IS-AUDIO header for server-side transcription
     */
    fun streamChatCompletionWithAudio(
        audioContent: String, // Base64 encoded WAV audio
        agent: P8Agent? = P8Agent.P8_SIMULATION,
        model: String = "gpt-4.1-mini"
    ): Flow<ChatStreamEvent> = flow {
        try {
            // Create request with audio content as user message
            val request = ChatCompletionRequest(
                model = model,
                messages = listOf(ChatMessage(MessageRole.USER, audioContent)),
                stream = true
            )
            
            val endpoint = when (agent) {
                null -> "${AppConfig.baseUrl}/api/v1/chat/completions"
                else -> "${AppConfig.baseUrl}/api/v1/agent/${agent.key}/chat/completions"
            }
            
            Log.d(TAG, "🎤 Starting audio chat stream to: $endpoint")
            
            val chatHeaders = headersManager.getChatHeaders(sessionId)
            Log.d(TAG, "🆔 Audio session ID: $sessionId")
            
            httpClient.preparePost(endpoint) {
                // Apply all standard headers
                chatHeaders.forEach { (key, value) ->
                    header(key, value)
                }
                
                // Add audio transcription header
                header("X-CHAT-IS-AUDIO", "true")
                
                contentType(ContentType.Application.Json)
                setBody(request)
            }.execute { response ->
                
                if (response.status != HttpStatusCode.OK) {
                    val errorText = response.bodyAsText()
                    Log.e(TAG, "❌ Audio chat failed: ${response.status} - $errorText")
                    emit(ChatStreamEvent.Error(Exception("Audio chat failed: ${response.status}")))
                    return@execute
                }
                
                Log.d(TAG, "🎤✅ Audio SSE stream connected")
                
                // Parse SSE stream (same logic as text chat)
                val channel = response.bodyAsChannel()
                val contentBuffer = StringBuilder()
                var doneEmitted = false
                
                while (!channel.isClosedForRead) {
                    val line = channel.readUTF8Line() ?: break
                    
                    if (line.startsWith("data: ")) {
                        val data = line.substring(6).trim()
                        
                        if (data == "[DONE]") {
                            Log.d(TAG, "🎤🏁 Audio stream completed")
                            emit(ChatStreamEvent.Done)
                            doneEmitted = true
                            break
                        }
                        
                        try {
                            val chunk = json.decodeFromString<ChatCompletionChunk>(data)
                            chunk.choices.firstOrNull()?.delta?.content?.let { content ->
                                emit(ChatStreamEvent.Content(content))
                                contentBuffer.append(content)
                            }
                        } catch (e: Exception) {
                            Log.w(TAG, "❌ Failed to parse audio stream chunk: $data", e)
                        }
                    }
                }
                
                if (!doneEmitted) {
                    emit(ChatStreamEvent.Done)
                }
            }
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ Audio streaming error", e)
            emit(ChatStreamEvent.Error(e))
        }
    }
    
    /**
     * Start a new chat conversation (new session ID)
     */
    fun startNewConversation(): String {
        val newSessionId = UUID.randomUUID().toString()
        Log.d(TAG, "🆕 Starting new conversation with session ID: $newSessionId")
        return newSessionId
    }
    
    fun close() {
        httpClient.close()
    }
}

/**
 * Headers manager for P8FS content headers
 * Implements all headers from content_headers.md
 */
class HeadersManager {
    
    fun getChatHeaders(sessionId: String): Map<String, String> {
        val headers = mutableMapOf<String, String>()
        
        // Authentication
        headers["Authorization"] = "Bearer ${getStoredAccessToken()}"
        
        // Session Management
        headers["X-Session-ID"] = sessionId
        headers["X-Tenant-ID"] = getStoredTenantId()
        
        // User Context
        headers["X-User-Email"] = getStoredUserEmail()
        headers["X-User-Source-ID"] = getStoredUserId()
        
        // Device Context
        headers["X-Device-ID"] = getDeviceId()
        headers["X-Device-Type"] = "mobile"
        headers["X-Device-Platform"] = "Android"
        headers["X-Device-Model"] = android.os.Build.MODEL
        headers["X-Device-Version"] = android.os.Build.VERSION.RELEASE
        headers["X-App-Version"] = getAppVersion()
        headers["X-OS-Version"] = android.os.Build.VERSION.RELEASE
        headers["X-Platform"] = "Android"
        
        // Mobile-specific
        headers["X-Biometric-Available"] = isBiometricAvailable().toString()
        headers["X-Secure-Enclave"] = hasSecureEnclave().toString()
        
        // Content Context for Chat
        headers["X-Content-Type"] = "application/json"
        headers["X-Content-Language"] = getDeviceLanguage()
        headers["X-Processing-Context"] = "chat-completion"
        
        Log.d("HeadersManager", "📋 Generated ${headers.size} chat headers")
        
        return headers
    }
    
    fun generateNewSessionId(): String = UUID.randomUUID().toString()
    
    // Implementation-specific methods
    private fun getStoredAccessToken(): String = TODO("Retrieve from secure storage")
    private fun getStoredTenantId(): String = TODO("Retrieve from secure storage") 
    private fun getStoredUserEmail(): String = TODO("Retrieve from secure storage")
    private fun getStoredUserId(): String = TODO("Retrieve from secure storage")
    private fun getDeviceId(): String = TODO("Get unique device identifier")
    private fun getAppVersion(): String = TODO("Get from BuildConfig")
    private fun isBiometricAvailable(): Boolean = TODO("Check biometric capability")
    private fun hasSecureEnclave(): Boolean = TODO("Check for secure hardware")
    private fun getDeviceLanguage(): String = TODO("Get device locale")
}

// Usage Example:
/*
class ChatViewModel : ViewModel() {
    
    private val headersManager = HeadersManager()
    private val chatClient = P8FSChatClient(headersManager)
    
    fun sendMessage(userMessage: String) {
        viewModelScope.launch {
            chatClient.streamChatCompletion(
                userMessage = userMessage,
                agent = P8Agent.P8_SIMULATION
            ).collect { event ->
                when (event) {
                    is ChatStreamEvent.Content -> {
                        // Update UI with streaming content
                        updateStreamingMessage(event.text)
                    }
                    is ChatStreamEvent.Done -> {
                        // Finalize message display
                        completeMessage()
                    }
                    is ChatStreamEvent.Error -> {
                        // Handle error
                        showError(event.exception.message ?: "Chat error")
                    }
                }
            }
        }
    }
    
    fun sendAudioMessage(audioData: String) {
        viewModelScope.launch {
            chatClient.streamChatCompletionWithAudio(
                audioContent = audioData,
                agent = P8Agent.P8_SIMULATION
            ).collect { event ->
                // Handle audio chat events same as text
            }
        }
    }
}
*/

File: docs/pseudo-code/sse-chat-parsing/SSEChatParser.swift
/**
 * P8FS SSE Chat Parsing Flow - iOS Implementation
 *
 * Based on reference Android implementation from T1 app
 * Key features:
 * 1. Session ID management for conversation threads
 * 2. Comprehensive header management from content_headers.md
 * 3. SSE streaming with URLSessionDataTask
 * 4. Only send latest message, server manages thread context
 * 5. Support for audio chat via X-CHAT-IS-AUDIO header
 */

import Foundation
import Combine
import os.log

// MARK: - Data Models

struct ChatMessage: Codable {
    let role: MessageRole
    let content: String
}

enum MessageRole: String, Codable {
    case user = "user"
    case assistant = "assistant"
}

struct ChatCompletionRequest: Codable {
    let model: String
    let messages: [ChatMessage]
    let temperature: Double
    let maxTokens: Int
    let stream: Bool
    
    enum CodingKeys: String, CodingKey {
        case model, messages, temperature, stream
        case maxTokens = "max_tokens"
    }
}

struct ChatCompletionChunk: Codable {
    let id: String
    let object: String
    let choices: [StreamChoice]
}

struct StreamChoice: Codable {
    let index: Int
    let delta: Delta
    let finishReason: String?
    
    enum CodingKeys: String, CodingKey {
        case index, delta
        case finishReason = "finish_reason"
    }
}

struct Delta: Codable {
    let content: String?
}

enum P8Agent: String, CaseIterable {
    case p8Simulation = "p8-sim"
    case p8Research = "p8-research" 
    case p8Analysis = "p8-analysis"
    case p8QA = "p8-qa"
}

enum ChatStreamEvent {
    case content(String)
    case done
    case error(Error)
}

// MARK: - Main Chat Client

@MainActor
class P8FSChatClient: ObservableObject {
    
    private let logger = Logger(subsystem: "app.p8fs", category: "ChatClient")
    private let headersManager: HeadersManager
    
    // Generate unique session ID for this chat conversation
    private let sessionId = UUID().uuidString
    
    private var urlSession: URLSession
    private var streamTask: URLSessionDataTask?
    
    init(headersManager: HeadersManager) {
        self.headersManager = headersManager
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 60.0 // 60 seconds for streaming
        config.timeoutIntervalForResource = 60.0
        
        self.urlSession = URLSession(configuration: config)
        
        logger.info("🆔 Generated new chat session ID: \(self.sessionId)")
    }
    
    /**
     * Stream chat completion with comprehensive P8FS headers
     * Only sends the latest user message - server manages thread via session ID
     */
    func streamChatCompletion(
        userMessage: String,
        agent: P8Agent? = .p8Simulation,
        model: String = "gpt-4.1-mini"
    ) -> AsyncThrowingStream<ChatStreamEvent, Error> {
        
        return AsyncThrowingStream { continuation in
            Task {
                do {
                    // Create request with only the new user message
                    let request = ChatCompletionRequest(
                        model: model,
                        messages: [ChatMessage(role: .user, content: userMessage)],
                        temperature: 0.7,
                        maxTokens: 1000,
                        stream: true
                    )
                    
                    // Determine endpoint based on agent
                    let baseURL = AppConfig.baseUrl
                    let endpoint: String
                    if let agent = agent {
                        endpoint = "\(baseURL)/api/v1/agent/\(agent.rawValue)/chat/completions"
                    } else {
                        endpoint = "\(baseURL)/api/v1/chat/completions"
                    }
                    
                    logger.info("🚀 Starting streaming chat to: \(endpoint)")
                    logger.info("📝 User message: \(userMessage)")
                    logger.info("🤖 Agent: \(agent?.rawValue ?? "none")")
                    
                    // Create URL request
                    guard let url = URL(string: endpoint) else {
                        throw ChatError.invalidURL
                    }
                    
                    var urlRequest = URLRequest(url: url)
                    urlRequest.httpMethod = "POST"
                    urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    
                    // Apply comprehensive P8FS headers
                    let chatHeaders = headersManager.getChatHeaders(sessionId: sessionId)
                    for (key, value) in chatHeaders {
                        urlRequest.setValue(value, forHTTPHeaderField: key)
                    }
                    
                    logger.info("🆔 Session ID: \(self.sessionId)")
                    logger.info("📋 Headers: \(chatHeaders.keys.joined(separator: ", "))")
                    
                    // Set request body
                    let jsonData = try JSONEncoder().encode(request)
                    urlRequest.httpBody = jsonData
                    
                    // Start streaming task
                    streamTask = urlSession.dataTask(with: urlRequest) { [weak self] data, response, error in
                        
                        if let error = error {
                            self?.logger.error("❌ Streaming error: \(error.localizedDescription)")
                            continuation.finish(throwing: error)
                            return
                        }
                        
                        guard let httpResponse = response as? HTTPURLResponse else {
                            continuation.finish(throwing: ChatError.invalidResponse)
                            return
                        }
                        
                        guard httpResponse.statusCode == 200 else {
                            self?.logger.error("❌ HTTP error: \(httpResponse.statusCode)")
                            continuation.finish(throwing: ChatError.httpError(httpResponse.statusCode))
                            return
                        }
                        
                        guard let data = data else { return }
                        
                        // Parse SSE stream
                        let dataString = String(data: data, encoding: .utf8) ?? ""
                        self?.parseSSEData(dataString, continuation: continuation)
                    }
                    
                    streamTask?.resume()
                    
                } catch {
                    logger.error("❌ Failed to start stream: \(error.localizedDescription)")
                    continuation.finish(throwing: error)
                }
            }
        }
    }
    
    /**
     * Stream chat with audio input
     * Sets X-CHAT-IS-AUDIO header for server-side transcription
     */
    func streamChatCompletionWithAudio(
        audioContent: String, // Base64 encoded WAV audio
        agent: P8Agent? = .p8Simulation,
        model: String = "gpt-4.1-mini"
    ) -> AsyncThrowingStream<ChatStreamEvent, Error> {
        
        return AsyncThrowingStream { continuation in
            Task {
                do {
                    let request = ChatCompletionRequest(
                        model: model,
                        messages: [ChatMessage(role: .user, content: audioContent)],
                        temperature: 0.7,
                        maxTokens: 1000,
                        stream: true
                    )
                    
                    let baseURL = AppConfig.baseUrl
                    let endpoint: String
                    if let agent = agent {
                        endpoint = "\(baseURL)/api/v1/agent/\(agent.rawValue)/chat/completions"
                    } else {
                        endpoint = "\(baseURL)/api/v1/chat/completions"
                    }
                    
                    logger.info("🎤 Starting audio chat stream to: \(endpoint)")
                    
                    guard let url = URL(string: endpoint) else {
                        throw ChatError.invalidURL
                    }
                    
                    var urlRequest = URLRequest(url: url)
                    urlRequest.httpMethod = "POST"
                    urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    
                    // Apply standard headers
                    let chatHeaders = headersManager.getChatHeaders(sessionId: sessionId)
                    for (key, value) in chatHeaders {
                        urlRequest.setValue(value, forHTTPHeaderField: key)
                    }
                    
                    // Add audio transcription header
                    urlRequest.setValue("true", forHTTPHeaderField: "X-CHAT-IS-AUDIO")
                    
                    logger.info("🆔 Audio session ID: \(self.sessionId)")
                    
                    let jsonData = try JSONEncoder().encode(request)
                    urlRequest.httpBody = jsonData
                    
                    streamTask = urlSession.dataTask(with: urlRequest) { [weak self] data, response, error in
                        
                        if let error = error {
                            self?.logger.error("❌ Audio streaming error: \(error.localizedDescription)")
                            continuation.finish(throwing: error)
                            return
                        }
                        
                        guard let httpResponse = response as? HTTPURLResponse,
                              httpResponse.statusCode == 200 else {
                            continuation.finish(throwing: ChatError.httpError(500))
                            return
                        }
                        
                        guard let data = data else { return }
                        
                        let dataString = String(data: data, encoding: .utf8) ?? ""
                        self?.parseSSEData(dataString, continuation: continuation)
                    }
                    
                    streamTask?.resume()
                    
                } catch {
                    logger.error("❌ Failed to start audio stream: \(error.localizedDescription)")
                    continuation.finish(throwing: error)
                }
            }
        }
    }
    
    /**
     * Parse Server-Sent Events data
     */
    private func parseSSEData(
        _ data: String, 
        continuation: AsyncThrowingStream<ChatStreamEvent, Error>.Continuation
    ) {
        let lines = data.components(separatedBy: .newlines)
        
        for line in lines {
            if line.hasPrefix("data: ") {
                let sseData = String(line.dropFirst(6)).trimmingCharacters(in: .whitespaces)
                
                if sseData == "[DONE]" {
                    logger.info("🏁 Stream completed with [DONE] signal")
                    continuation.yield(.done)
                    continuation.finish()
                    return
                }
                
                // Parse JSON chunk
                if let jsonData = sseData.data(using: .utf8) {
                    do {
                        let chunk = try JSONDecoder().decode(ChatCompletionChunk.self, from: jsonData)
                        
                        if let content = chunk.choices.first?.delta.content {
                            logger.debug("📝 Received chunk: '\(content)'")
                            continuation.yield(.content(content))
                        }
                        
                        if let finishReason = chunk.choices.first?.finishReason {
                            logger.info("🏁 Finish reason: \(finishReason)")
                        }
                        
                    } catch {
                        logger.error("❌ Error parsing chunk: \(sseData) - \(error.localizedDescription)")
                    }
                }
            }
        }
    }
    
    /**
     * Start a new chat conversation (new session ID)
     */
    func startNewConversation() -> String {
        let newSessionId = UUID().uuidString
        logger.info("🆕 Starting new conversation with session ID: \(newSessionId)")
        return newSessionId
    }
    
    /**
     * Cancel current streaming task
     */
    func cancelStream() {
        streamTask?.cancel()
        streamTask = nil
    }
}

// MARK: - Headers Manager

class HeadersManager {
    
    private let logger = Logger(subsystem: "app.p8fs", category: "HeadersManager")
    
    func getChatHeaders(sessionId: String) -> [String: String] {
        var headers: [String: String] = [:]
        
        // Authentication
        headers["Authorization"]  = "Bearer \(getStoredAccessToken())"
        
        // Session Management
        headers["X-Session-ID"] = sessionId
        headers["X-Tenant-ID"] = getStoredTenantId()
        
        // User Context
        headers["X-User-Email"] = getStoredUserEmail()
        headers["X-User-Source-ID"] = getStoredUserId()
        
        // Device Context
        headers["X-Device-ID"] = getDeviceId()
        headers["X-Device-Type"] = "mobile"
        headers["X-Device-Platform"] = "iOS"
        headers["X-Device-Model"] = UIDevice.current.model
        headers["X-Device-Version"] = UIDevice.current.systemVersion
        headers["X-App-Version"] = getAppVersion()
        headers["X-OS-Version"] = UIDevice.current.systemVersion
        headers["X-Platform"] = "iOS"
        
        // iOS-specific
        headers["X-Biometric-Available"] = String(isBiometricAvailable())
        headers["X-Secure-Enclave"] = String(hasSecureEnclave())
        
        // Content Context for Chat
        headers["X-Content-Type"] = "application/json"
        headers["X-Content-Language"] = getDeviceLanguage()
        headers["X-Processing-Context"] = "chat-completion"
        
        logger.info("📋 Generated \(headers.count) chat headers")
        
        return headers
    }
    
    func generateNewSessionId() -> String {
        return UUID().uuidString
    }
    
    // Implementation-specific methods
    private func getStoredAccessToken() -> String {
        // TODO: Retrieve from Keychain
        return ""
    }
    
    private func getStoredTenantId() -> String {
        // TODO: Retrieve from secure storage
        return ""
    }
    
    private func getStoredUserEmail() -> String {
        // TODO: Retrieve from secure storage
        return ""
    }
    
    private func getStoredUserId() -> String {
        // TODO: Retrieve from secure storage
        return ""
    }
    
    private func getDeviceId() -> String {
        // TODO: Get unique device identifier
        return UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
    }
    
    private func getAppVersion() -> String {
        return Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "1.0"
    }
    
    private func isBiometricAvailable() -> Bool {
        // TODO: Check biometric capability with LocalAuthentication
        return false
    }
    
    private func hasSecureEnclave() -> Bool {
        // TODO: Check for Secure Enclave availability
        return false
    }
    
    private func getDeviceLanguage() -> String {
        return Locale.preferredLanguages.first ?? "en-US"
    }
}

// MARK: - Error Types

enum ChatError: LocalizedError {
    case invalidURL
    case invalidResponse
    case httpError(Int)
    case parsingError
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid response"
        case .httpError(let code):
            return "HTTP error: \(code)"
        case .parsingError:
            return "Parsing error"
        }
    }
}

// MARK: - App Config

struct AppConfig {
    static let baseUrl = "https://p8fs.percolationlabs.ai"
}

// Usage Example:
/*
@MainActor
class ChatViewModel: ObservableObject {
    
    @Published var messages: [ChatMessage] = []
    @Published var isStreaming = false
    @Published var currentStreamingContent = ""
    
    private let headersManager = HeadersManager()
    private let chatClient: P8FSChatClient
    
    init() {
        self.chatClient = P8FSChatClient(headersManager: headersManager)
    }
    
    func sendMessage(_ userMessage: String) {
        // Add user message immediately
        messages.append(ChatMessage(role: .user, content: userMessage))
        
        // Start streaming assistant response
        isStreaming = true
        currentStreamingContent = ""
        
        Task {
            do {
                for try await event in chatClient.streamChatCompletion(
                    userMessage: userMessage,
                    agent: .p8Simulation
                ) {
                    await MainActor.run {
                        switch event {
                        case .content(let text):
                            currentStreamingContent += text
                        case .done:
                            messages.append(ChatMessage(role: .assistant, content: currentStreamingContent))
                            isStreaming = false
                            currentStreamingContent = ""
                        case .error(let error):
                            print("Chat error: \(error.localizedDescription)")
                            isStreaming = false
                        }
                    }
                }
            } catch {
                await MainActor.run {
                    print("Stream error: \(error.localizedDescription)")
                    isStreaming = false
                }
            }
        }
    }
    
    func sendAudioMessage(_ audioData: String) {
        Task {
            for try await event in chatClient.streamChatCompletionWithAudio(
                audioContent: audioData,
                agent: .p8Simulation
            ) {
                // Handle audio chat events same as text
                await MainActor.run {
                    switch event {
                    case .content(let text):
                        currentStreamingContent += text
                    case .done:
                        messages.append(ChatMessage(role: .assistant, content: currentStreamingContent))
                        isStreaming = false
                        currentStreamingContent = ""
                    case .error(let error):
                        print("Audio chat error: \(error.localizedDescription)")
                        isStreaming = false
                    }
                }
            }
        }
    }
}
*/